DOC: Update release instructions

pllim opened the pull request:
Fix #5120 .

I am sure the first pass missed something. Please review carefully.

pllim commented on docs/development/releasing.rst:
```diff
@@ -299,29 +306,42 @@ packages that use the full bugfix/maintenance branch approach.)
 
+#. Send out release annoucement emails to email lists in
+   `Getting Help with Astropy <https://www.astropy.org/help.html>`_. Use the
+   previous annoucement as a template.
```
Ops, typo! Will fold this into the next edit cycle.
```suggestion
   previous announcement as a template.
```

pllim commented on docs/development/releasing.rst:
```diff
@@ -299,29 +306,42 @@ packages that use the full bugfix/maintenance branch approach.)
-     confuse some version number parsing tools.
-   * Do not do step #26 or later, as those are tasks for an actual release.
-
+For major releases we do beta and/or release candidates to have a chance to
```
Not sure why it was indented, so I simply un-indented the existing text. Then, added some.

astrofrog replied:
Seems reasonable

pllim commented on docs/development/releasing.rst:
```diff
@@ -299,29 +306,42 @@ packages that use the full bugfix/maintenance branch approach.)
+  If an RC goes well, there's no need for a "dev" stage, as the same version
+  will be released with only minor doc updates, and strings like "x.yrcz.dev"
+  confuse some version number parsing tools.
+* Do not do step #27 (the one *after* "Update Readthedocs...") or later, as those are tasks for an actual release.
```
I interpret the old comments as best I can... but the number might be off. :woman_shrugging:

astrofrog replied:
Maybe those tasks should be in their own section so that you can refer to the section label instead?

pllim replied:
Done...

pllim commented:
@lglattly , do you want to join the party too? If not, we can wait for the copy-editing until this gets on your list.

bsipocz commented on docs/development/releasing.rst:
```diff
@@ -148,20 +153,11 @@ packages that use the full bugfix/maintenance branch approach.)
 
       $ conda create -n astropy_release_test_v<version> numpy
       $ source activate astropy_release_test_v<version>
+      $ pip install .[all]
```
this would also install astropy, from the local directory, doesn't it?

pllim replied:
No idea. I haven't use `pip` like that much yet. Should I add a `pip uninstall astropy` after? Or should I find another way to "install all the dependencies" like @astrofrog wanted in #5120 ?

astrofrog replied:
A cleaner approach is to combine this with the line below:

```
pip install dist/astropy-<version>.tar.gz[all]
```

bsipocz commented on docs/development/releasing.rst:
```diff
@@ -236,6 +233,16 @@ packages that use the full bugfix/maintenance branch approach.)
       $ git reset --hard v<version>
       $ git push upstream stable --force
 
+#. Build and test the Astropy wheels.  See the `wheel builder README
```
nope, wheel building is before pushing to pypi. there has been sometimes issues uncovered by it, so I'm uncomfortable doing anything like this after pypi release.

pllim replied:
I moved it here because of the comment of "moving the wheel stuff after the tagging". Did I move it too far down? The numbering doesn't match anymore between the original issue and the doc, so I am just guessing here.

bsipocz replied:
yes, tagging, wheels and the very last step is pypi upload (and then conda-forge).

pllim replied:
OK moving it back

bsipocz commented on docs/development/releasing.rst:
```diff
@@ -396,6 +416,12 @@ The procedure for this is straightforward:
 
 #. Repeat the above steps for the astropy-helpers, using the same version series.
 
+#. Create an issue titled "Release final vX.Y".
```
does this issue needed?

pllim replied:
@astrofrog wanted it. 🤷‍♀

bsipocz replied:
oh, sure. but that was 3 years ago and he didn't complained since of not doing it. I'm OK either way, but feel it's superfluous, we never forgot to release even without having an issue.

pllim replied:
Tom R, what say you?

astrofrog replied:
Remove it! 🔥

bsipocz commented on docs/development/releasing.rst:
```diff
@@ -396,6 +416,12 @@ The procedure for this is straightforward:
 
+#. Create an issue titled "Release final vX.Y".
+
+#. Ping people to test it under some unusual cases like 32-bit Linux systems.
```
does this needed? the release candidate announcement was usually picked up by the right people (debian and st packaging)

pllim replied:
Just following check box written by @astrofrog .

astrofrog replied:
Yeah I don't think this is needed anymore as the Debian folk follow our list.

bsipocz commented:
some comments first, I haven't looked into what was in the original issue yet.

pllim commented:
FWIW, I think the 4.0 release would be a good exercise to try out these modified instructions and finally close #5120. 😉

bsipocz commented:
> FWIW, I think the 4.0 release would be a good exercise to try out these modified instructions and finally close

sounds good. Though I hardly ever follow the instructions, it's already in muscle memory 😄

lglattly commented:
I can join the party now! Just let me know when comments/changes have been implemented and I'll take a look!

astrofrog commented on docs/development/releasing.rst:
```diff
@@ -446,6 +472,14 @@ be backported.  For a more detailed set of guidelines on using milestones, see
+.. note::
+
+    The changelog script in ``astropy-procedures`` does not know about
+    minor releases, thus please becareful. For example, if PRs milestoned
```
typo: becareful

pllim commented on docs/development/releasing.rst:
```diff
@@ -228,6 +234,11 @@ packages that use the full bugfix/maintenance branch approach.)
 
+.. _post-release-procedure:
+
+Post-Release procedures
```
Is this where I should draw the line between "release" and "post-release"?

bsipocz replied:
pypi upload? after that the release cannot be modified? But honestly I don't really know.

pllim replied:
Okay. I moved post-release section to right after `twine upload`. The significance of this section separation is that in the RC instructions, it says "do not do the stuff in post-release".

pllim commented on docs/development/releasing.rst:
```diff
@@ -299,29 +310,42 @@ packages that use the full bugfix/maintenance branch approach.)
+  If an RC goes well, there's no need for a "dev" stage, as the same version
+  will be released with only minor doc updates, and strings like "x.yrcz.dev"
+  confuse some version number parsing tools.
+* Do not do steps in :ref:`post-release-procedure`.
```
Please check that the modified modifications still make sense.

pllim commented on docs/development/releasing.rst:
```diff
@@ -396,6 +420,8 @@ The procedure for this is straightforward:
 
 #. Repeat the above steps for the astropy-helpers, using the same version series.
 
+#. Run the tests twice in a row without cleaning anything up in between.
```
I removed 32-bit testing from here, but you still want to keep this, right?

bsipocz replied:
sure 🤷‍♀ 

(confession: most of the time I don't run them at this point any more as they take forever and in "CI we trust" 🙈 )

pllim replied:
@astrofrog , you requested this. What say you?

pllim replied:
@astrofrog and I agreed that this can be removed from release instruction and added to CI directly at a later time.

lglattly commented on docs/development/releasing.rst:
```diff
@@ -228,6 +234,11 @@ packages that use the full bugfix/maintenance branch approach.)
 
+.. _post-release-procedure:
+
+Post-Release procedures
```
```suggestion
Post-Release Procedures
```

pllim replied:
Not sure about this one, as the other existing headings only capitalized the first letter of the first word.

lglattly replied:
No worries, will edit all of the headings when I get to this sub-package!

lglattly commented on docs/development/releasing.rst:
```diff
@@ -299,29 +310,42 @@ packages that use the full bugfix/maintenance branch approach.)
-     confuse some version number parsing tools.
-   * Do not do step #26 or later, as those are tasks for an actual release.
-
+For major releases we do beta and/or release candidates to have a chance to
```
```suggestion
For major releases, we do beta and/or release candidates to have a chance to
```

lglattly commented on docs/development/releasing.rst:
```diff
@@ -299,29 +310,42 @@ packages that use the full bugfix/maintenance branch approach.)
+  ``.dev``, enter "1.2b1" or "1.2rc1".  It is critical that you follow this
+  numbering scheme (``x.yb#`` or ``x.y.zrc#``), as it will ensure the release
+  is ordered "before" the main release by various automated tools, and also
+  tells PyPI that this is a "pre-release".
```
```suggestion
  tells PyPI that this is a "pre-release."
```

lglattly commented on docs/development/releasing.rst:
```diff
@@ -299,29 +310,42 @@ packages that use the full bugfix/maintenance branch approach.)
+  is ordered "before" the main release by various automated tools, and also
+  tells PyPI that this is a "pre-release".
+* Do *not* do the step of adding ``.dev`` in the "back to development" stage.
+  If an RC goes well, there's no need for a "dev" stage, as the same version
```
```suggestion
  If an RC goes well, there is no need for a "dev" stage, as the same version
```

lglattly commented on docs/development/releasing.rst:
```diff
@@ -299,29 +310,42 @@ packages that use the full bugfix/maintenance branch approach.)
+  confuse some version number parsing tools.
+* Do not do steps in :ref:`post-release-procedure`.
+
+Once a release candidate is available, create a new wiki page under
```
```suggestion
Once a release candidate is available, create a new Wiki page under
```

pllim commented on docs/development/releasing.rst:
```diff
@@ -89,6 +84,16 @@ packages that use the full bugfix/maintenance branch approach.)
+   for the `astropy core repository`_ branch you are going to release. You may
+   also want to locally run the tests (with remote data on to ensure all of the
+   tests actually run), and make sure the description in ``setup.cfg`` is reStructuredText-
+  compliant::
```
Ops... I think it was underindented. Will fix in the commit.
```suggestion
   compliant::
```

lglattly replied:
oops I meant to make reStructureText-compliant one hyphenated word, but wasn't sure how to do that here.

pllim replied:
That is a limitation of this "suggestion" feature on GitHub. It does not do well with multi-line editing. But no matter, I'll sort this out in my next batch of edits. Thank you!

eteq commented on docs/development/releasing.rst:
```diff
@@ -89,6 +84,16 @@ packages that use the full bugfix/maintenance branch approach.)
 
       $ git checkout v1.2.x
 
+#. Make sure that the continuous integration services (e.g., Travis) are passing
```
Perhaps
```suggestion
#. Make sure that the continuous integration services (e.g., CircleCI) are passing
```
just because of all the uncertainty in Travis' future?

bsipocz replied:
What about listing multiple? I regularly point to this docs for folks looking for instructions, and in the template we provide more for travis than for circleCI, thus many projects only use that one.

pllim replied:
OK, I added both.

eteq commented on docs/development/releasing.rst:
```diff
@@ -299,29 +310,42 @@ packages that use the full bugfix/maintenance branch approach.)
+`wiki of a previous RC <https://github.com/astropy/astropy/wiki/v3.2-RC-testing>`_
+as a template.
+
+For releases that come after release candidates, the title of the changelog
```
It's a bit awkward, but I think this goes in the "main" release instructions, right?  Because this section is for "beta/RC" but this note is for *non*-beta/RC releases

pllim replied:
I moved it up but you can tell me whether it is where you thought it is supposed to be. 🤞

bsipocz commented:
OK. Post release procedure seems legit. I may have some updates to the order of a few steps after doing the procedure again, e.g. labeling/milestone updates and the changelog cleanups might just need to come a bit earlier, in fact they are not much less part of the procedure (e.g. labels and milestones are done much earlier, while the changelog changes needs to be in a PR as only the 4 CoCo members have direct push access to master.)

bsipocz commented:
OK, so while doing LTS release these are the remaining comments:

* I don't do 3), but it's already flagged as optional
* I have two bugfix branches as worktrees. However worktree doesn't work well with submodules, so to do the sdist I go back to the main checkout I have. This doesn't listed anywhere, but if you like it can be mentioned.
* I don't do 15&16), by the time I'm there in the list the test suite has been run quite a lot of times everywhere, whenever I did this step in the past  never recovered any new issues that didn't come up on CI or with the wheels building.
* building the wheels (with MacPython) requires a tag available on github, so 17) cannot be done before pushing the tag to github. then 18) might need to be done as well as a force push of the tag to github.

post reelase:
* 3) should be done earlier to get the tag available for the wheels
* 3) `git push upstream v<version branch>.x` needs to  be done again here with the new changelog/version info
 * 8) cannot be done as described, changelog has to be in a PR as `master` now a protected branch.
* 9) 10) - I do these much earlier than the release. Clearing out the milestone is basically my first step to start the release (after nagging folks to finish up PRs). I usually close the milestone when I start the release process. Milestones for new bugfix releases are created when the first call for finishing up stuff goes out (as remilestoning is a perfectly valid maintainer response for that call).
*11) `conda-forge` bot opens a PR automatically for stable releases, but for LTS it has to be manual against the lts branch (same with the wheels).

Other practices:
* Always start with LTS release then stable bugfix. Do them in pairs, or LTS only otherwise there will be a timewarp with the changelog and what fix got included in which version.
* I regularly diff the backlog changes to master changes in case something diverged (the scripts cannot handle the backport one, only works with master). Then, after the release I just copy paste the new section overwriting whatever exists in master.
* `add_to_changelog.py` never worked for me, I just copy paste from the new sections from the same commit from the previous release (and remove if modules got removed/added). It's quick thus I didn't bothered with fixing the script.
* we need to update of what should happen with the what's news. Also, in the announcement, the links shouldn't point to `stable` rather to the release tag.

pllim commented:
@bsipocz , thanks for the valuable input! I think I addressed most of your feedback, except for these:

> but for LTS it has to be manual against the lts branch (same with the wheels)

I don't do conda release, so I am not sure what that means. Also, seems like the instructions on this page gloss over things like how to do wheels or conda-forge, so I just left this detail out.

> I regularly diff the backlog changes to master changes in case something diverged (the scripts cannot handle the backport one, only works with master). Then, after the release I just copy paste the new section overwriting whatever exists in master.

I don't understand this enough to know where or how to put it in words that are consistent with the narrative.

> we need to update of what should happen with the what's news. 

Could you please elaborate?

bsipocz commented:
> I don't do conda release, so I am not sure what that means. Also, seems like the instructions on this page gloss over things like how to do wheels or conda-forge, so I just left this detail out.

For conda-forge, their bot opens PRs to the latest pypi releases, so usually there is nothing to do other than merge that (or follow the instructions within the PR).
For LTS however they don't pick up versions, so one needs to open a PR against the lts branch here: https://github.com/conda-forge/astropy-feedstock/

> > I regularly diff the backlog changes to master changes in case something diverged (the scripts cannot handle the backport one, only works with master). Then, after the release I just copy paste the new section overwriting whatever exists in master.
> 
> I don't understand this enough to know where or how to put it in words that are consistent with the narrative.

this is one of those that is convenient for me and very different from what was convenient for Erik. So it's fine to keep what we have now as who knows that's convenient for any third party.

> > we need to update of what should happen with the what's news.

We haven't yet included here that for any what's new that's has a branch, the content of the file on master should be removed and replaced with a link to the branch (or remove the file altogether and do the release announcement e-mail with the version link rather than with one that point to "stable".)

pllim commented on docs/development/releasing.rst:
```diff
@@ -446,6 +496,14 @@ be backported.  For a more detailed set of guidelines on using milestones, see
 
+.. note::
+
+    The changelog script in ``astropy-procedures`` does not know about
```
I removed all mentions of using `add_to_changelog.py` since @bsipocz that the script does not work anyway. Does that mean I should also remove this note, or is this referring to a different "changelog script"?

bsipocz replied:
no, this comment is about the scripts in `pr_consistency`, they only check for a PR being backported to a branch, rather than checking for the milestones, too. As the comment says we have two branches: 1) master 2) 1.2.x. The release 1.2.0 will come out of 1.2.x, just as well as 1.2.1. If a PR is merged into master with 1.2.1 it shouldn't be backported until 1.2.0 is out, but the script will complain that it needs to be backported to 1.2.x nevertheless.

bsipocz commented on docs/development/releasing.rst:
```diff
@@ -299,21 +299,26 @@ Post-Release procedures
+   "what's new" documentation in the released RTD branch, using the existing
+   text as example.
+
+#. Notify the Conda Distribution and Continuous Integration maintainers
```
Oh, I haven't noticed this before. Remove the first sentence, and add a new point after this one: 

* "When the conda-forge packaging is ready, email the Anaconda maintainers about the release(s) so they can update the versions in the default channels."

* "Once the release(s) are available on the default conda channels, prepare the public announcement. For new major release, coordinate with the astropy coordinators, for bugfix release(s) an e-mail to the mailing lists `astropy-dev`` and ``Astropy`` is sufficient."

pllim replied:
Isn't the second point you mentioned already the last point in this section, with "Send out release announcement emails to email lists..."?

bsipocz replied:
yes, that's right. it's just adding a tiny nuance that the announcement is expected to be coordinated, and not sent out by the release manager but by the coordinators.

pllim replied:
Anyway, I attempted to apply your comments to where I think is appropriate. :woman_shrugging:

bsipocz commented:
I'm sorry @pllim to be a pain with this, there was reasons that this all lived in a never ending wishlist in an issue 😅

bsipocz commented:
> I'm sorry @pllim to be a pain with this, there was reasons that this all lived in a never ending wishlist in an issue

so basically please use your power to say "that's enough" at any point and merge this. It's already infinitely better than before, and we can always go back to said issue and add more to the list as they come up.

pllim commented:
@bsipocz , no worries! You and @eteq should have another read through just in case. Otherwise, I think I have addressed all the comments.

bsipocz commented:
Try to get hold of him then, I'm happy with this now.

eteq commented on docs/development/releasing.rst:
```diff
@@ -197,8 +245,14 @@ packages that use the full bugfix/maintenance branch approach.)
+      $ twine check dist/*
       $ twine upload dist/astropy-<version>*
 
+.. _post-release-procedure:
```
```suggestion

Congratulations!  You have completed the release! Now there are just a few
clean-up tasks to finalize the process.

.. _post-release-procedure:
```

bsipocz replied:
Is this sarcasm? 😏

pllim replied:
Wait, am I supposed to commit that in or not, because I did. :scream:

bsipocz replied:
It wasn't intended to be sarcasm, I think, but usually the most tedious, less fun part is the cleanups, updating the contributors page and mailmaps, etc. 🙈
